### ISSUE TITLE
Using own credentials requires re-auth every few days

### ISSUE THREAD

**Issue by erhe** [sabeechen/hassio-google-drive-backup#347]
Using the addon and using own credentials against Google Drive, works well, except I need to re-auth every few days.
The logs are showing the following when it occurs.

######  Log output
 ```
02-11 13:54:57 INFO [backup.model.coordinator] Syncing Snapshots
02-11 13:54:57 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
02-11 13:54:57 DEBUG [backup.drive.driverequests] Requesting refreshed Google Drive credentials
02-11 13:54:57 ERROR [backup.model.coordinator] Couldn't refresh your credentials with Google because: 'invalid_grant'
02-11 13:54:57 INFO [backup.model.coordinator] I'll try again in an hour
 ```

Opening the console and copying in the client id/secret and adding the token back to the addon allows it to work for a few days.

###### Description:

 Addon version: `0.103.1`
 Home Assistant Version: `2021.2.2`
 Supervisor Version: `2021.02.6`
 Supervisor Channel: `stable`
 Hassos Version: `None`
 Docker Version: `19.03.14`
 Architecture: `amd64`
 Machine: `qemux86-64`
 Date: `2021-02-11T13:44:13.556676+00:00`
 Timezone: `Europe/Stockholm`
 Failure Time: `2021-02-11T12:54:57.633329+00:00`
 Last Good Sync: `2021-02-11T13:32:17.018504+00:00`
 ###### Exception:
 ```
 No error could be identified automatically.
 ```
 Snapshots:
 ```
 {
    "HomeAssistant": {
        "snapshots": 22,
        "retained": 2,
        "deletable": 20,
        "name": "HomeAssistant",
        "title": "Home Assistant",
        "latest": "2021-02-11T13:32:04Z",
        "max": 20,
        "enabled": true,
        "size": "457.1 MB",
        "free_space": "44.0 GB"
    },
    "GoogleDrive": {
        "snapshots": 59,
        "retained": 1,
        "deletable": 58,
        "name": "GoogleDrive",
        "title": "Google Drive",
        "latest": "2021-02-11T13:32:04Z",
        "max": 60,
        "enabled": true,
        "size": "1.8 GB"
    }
}
 ```
 ###### Config:
 ```
 {
    "Setting.MAX_SNAPSHOTS_IN_HASSIO": 20,
    "Setting.MAX_SNAPSHOTS_IN_GOOGLE_DRIVE": 60,
    "Setting.DAYS_BETWEEN_SNAPSHOTS": 1.0
}
 ```

**Comment by sabeechen** [comment]
Sorry this took me a while to get back to you on, I had to step away from staying up to date on issues for a little while to take care of some personal stuff.

I'd usually recommend that you re-create new credentials, but it looks like Google changed the workflow for making credentials in little ways and now my instructions are out of date.  There are a lot of little mistakes someone can make and I'll need to do some testing to make sure things still work.  Unfortunately thats going to mean I make new creds and just wait to see if the error comes up.  I'll get back to you in a few days and see if I can reproduce this problem with the creds I just made.

**Comment by netmikey** [comment]
I'm having the exact same issue here. I also just set this up about a week ago so your guess that it might have something to do with changes on Google's end seems plausible. After, well, googling a bit, I found this blog entry which seems to be a very good checklist for this particular issue. Let me link it here just in case you haven't seen it already:

https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35

**Comment by sabeechen** [comment]
Ah, that explains it.  The "Testing" status wasn't a thing when I first made the instructions.  I'm not sure when it was added.

I was able to make "Testing" credentials in Google's Cloud console (which is what you should already have) and then go to [https://console.cloud.google.com/apis/credentials/consent](https://console.cloud.google.com/apis/credentials/consent) and click "Publish", which takes them out of testing mode.  Google didn't require me to submit the Oauth consent screen for verification, it seems to have taken effect immediately.  I believe this is because the permission the addon requires (drive.file aka "Only access the files it creates") is not considered a *sensitive* scope.

Could I have you guys click that "Publish" button and then re-authorize the addon?  I strongly suspect that will resolve this and if so I'll add it to the instructions.  I don't think we'll know for 7 days or so.

Side note: I've never fully understood how Google's Oauth consent screen verification works.  I suspects its intentionally opaque.  The credentials most users use to authorize with Google Drive (the one that takes them to habackup.io to login) is still listed as "Pending Verification" and has been that way for 2 years, but hasn't caused any problems yet.  I'm afraid to touch it for fear of it breaking everyone everyone :)

**Comment by sabeechen** [comment]
Also just to reiterate, I think you'll need to reauthorize the addon (eg enter your client secret and client id into the addon again) since the prior grant Google gave it was for the credentials in "testing" mode, which has the shorter 7-day expiration date.

**Comment by netmikey** [comment]
I can confirm that publishing didn't require verification. Mine is "In production" status now and I have reauthorized the addon. I'll report here if it still works in > 7 days :) (feel free to ping me as of 01.03.2021 if I forget to... ;-))

Thanks for your support, your time, and this awesome add-on, @sabeechen !

**Comment by sabeechen** [comment]
/remind me about this issue in 8 days

Me too, so this is a great opportunity to test out the reminder bot.

**Comment by erhe** [comment]
> Sorry this took me a while to get back to you on, I had to step away from staying up to date on issues for a little while to take care of some personal stuff.
> 
> I'd usually recommend that you re-create new credentials, but it looks like Google changed the workflow for making credentials in little ways and now my instructions are out of date. There are a lot of little mistakes someone can make and I'll need to do some testing to make sure things still work. Unfortunately thats going to mean I make new creds and just wait to see if the error comes up. I'll get back to you in a few days and see if I can reproduce this problem with the creds I just made.

No problem,

**Comment by erhe** [comment]
Thank you for getting back, mine was indeed in Testing, makes sense, I have published it and we will see next week.
Cheers!

**Comment by sabeechen** [comment]
Did any of ya'lls credentials expire?  Mine seem to still be working, so I'm tempted to say this is solution.  I'll update the documentation if its working for you guys too.

**Comment by erhe** [comment]
> Did any of ya'lls credentials expire? Mine seem to still be working, so I'm tempted to say this is solution. I'll update the documentation if its working for you guys too.

Mine works well, no issues since publishing the app.

**Comment by netmikey** [comment]
I can confirm: still working here too 👍

**Comment by sabeechen** [comment]
Great

**Comment by glyph-se** [comment]
Hi,

I got the same problem and also managed to get this working by putting my app in production, no verification needed. Not sure if you saw it, but I saw this text while doing so:

>  Your app will be available to any user with a Google Account.
>
> You do not need to submit your app for verification. If you change your app's configuration in the future, like adding more than 10 domains, uploading a logo, or requesting sensitive or restricted scopes, you will need to submit for verification. 

I had to re-authenticate the addon, but could use the same oauth client id and secret.

**Comment by sabeechen** [comment]
I just released the new auth instructions with v 0.104.1 of the addon, let me know if you guys run into any more problems.

To update the addon, go to:
`Supervisor` > `Add-on Store` > Click the `...` in the top right > `Refresh`
and you should see the new version.
